jerry-core/ecma/builtin-objects/ecma-builtin-object.c

```diff
@@ -295,57 +258,181 @@ ecma_builtin_object_object_get_own_property_symbols (ecma_object_t *obj_p) /**<
 #endif /* ENABLED (JERRY_ES2015) */
 
 /**
- * The Object object's 'seal' routine
+ * SetIntegrityLevel operation
  *
  * See also:
- *          ECMA-262 v5, 15.2.3.8
+ *          ECMA-262 v6, 7.3.14
  *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
+ * @return ECMA_VALUE_ERROR - if the operation raised an error
+ *         ECMA_VALUE_{TRUE/FALSE} - depends on whether the integrity level has been set sucessfully
  */
 static ecma_value_t
-ecma_builtin_object_object_seal (ecma_object_t *obj_p) /**< routine's argument */
+ecma_builtin_object_set_integrity_level (ecma_object_t *obj_p, /**< object */
+                                         bool is_seal) /**< true - "sealed" level
+                                                        *   false - "frozen" level */
 {
+  /* 3. */
+#if ENABLED (JERRY_ES2015_BUILTIN_PROXY)
+  if (ecma_object_is_proxy (obj_p))
+  {
+    ecma_value_t status = ecma_proxy_object_prevent_extensions (obj_p);
+
+    if (ECMA_IS_VALUE_ERROR (status))
+    {
+      return status;
+    }
+
+    if (ecma_is_value_false (status))
+    {
+      return status;
+    }
+
+    JERRY_ASSERT (ecma_is_value_true (status));
+  }
+  else
+  {
+#endif /* ENABLED (JERRY_ES2015_BUILTIN_PROXY) */
+    ecma_op_ordinary_object_prevent_extensions (obj_p);
+#if ENABLED (JERRY_ES2015_BUILTIN_PROXY)
+  }
+#endif /* ENABLED (JERRY_ES2015_BUILTIN_PROXY) */
+
+  /* 6. */
   ecma_collection_t *props_p = ecma_op_object_get_property_names (obj_p, ECMA_LIST_CONVERT_FAST_ARRAYS);
 
+#if ENABLED (JERRY_ES2015_BUILTIN_PROXY)
+  if (props_p == NULL)
+  {
+    return ECMA_VALUE_ERROR;
+  }
+#endif /* ENABLED (JERRY_ES2015_BUILTIN_PROXY) */
+
   ecma_value_t *buffer_p = props_p->buffer_p;
 
-  for (uint32_t i = 0; i < props_p->item_count; i++)
+  if (is_seal)
```

Comment:
Both parts of this if is very similar (I've spotted only 4-5 lines of difference, a single if). Based on the standard however there should be more differences. Is this correct? If it is , can we merge the two sides of the if somehow?

Reply:
According to the ES5 spec they are really similar, but in ES6 there are significant differences. This should be updated, but out of this patch scope.